examples/with-mongodb-mongoose/utils/dbConnect.js

```diff
@@ -1,18 +1,45 @@
 import mongoose from 'mongoose'
+  
+const { MONGODB_URI } = process.env
+
+if (!MONGODB_URI) {
+  throw new Error(
+    'Please define the MONGODB_URI environment variable inside .env.local'
+  )
+}
+
+/**
+ * Global is used here to maintain a cached connection across hot reloads
+ * in development. This prevents connections growing exponentially
+ * during API Route usage.
+ */
+let cached = global.mongoose
+
+if (!cached) {
+  cached = global.mongoose = { conn: null, promise: null }
+}
 
 async function dbConnect() {
-  // check if we have a connection to the database or if it's currently
-  // connecting or disconnecting (readyState 1, 2 and 3)
-  if (mongoose.connection.readyState >= 1) {
-    return
+  if (cached.conn) {
+    return cached.conn
   }
 
-  return mongoose.connect(process.env.MONGODB_URI, {
-    useNewUrlParser: true,
-    useUnifiedTopology: true,
-    useFindAndModify: false,
-    useCreateIndex: true,
-  })
+  if (!cached.promise) {
+    const opts = {
+      useNewUrlParser: true,
+      useUnifiedTopology: true,
+      bufferCommands: false,
+      bufferMaxEntries: 0,
+      useFindAndModify: false,
+      useCreateIndex: true,
+    }
+
+    cached.promise = mongoose.connect(MONGODB_URI, opts).then((mongoose) => {
+      return mongoose
+    })
```

Comment:
Usually mongoose is used like this:

```js
import mongoose from 'mongoose'

mongoose.//
```

The first time the connection is created it will be fine, but when re-used the mongoose connections array won't be touched, and instead it will work if used directly from the result of this function call.

Let's make it so `mongoose.connections` is updated if the connection is re-used.

Reply:
After testing the changes here I can't see an important increase on the number of connections while testing different API routes and pages. So I don't think changing to `mongoose.createConnection` is required :thinking: